bioptim/dynamics/dynamics_functions.py

```diff
@@ -687,6 +805,41 @@ def dispatch_q_qdot_tau_data(states: MX.sym, controls: MX.sym, nlp) -> tuple:
         tau = nlp.mapping["tau"].to_second.map(controls[: nlp.shape["tau"]])
         return q, qdot, tau
 
+    @staticmethod
+    def dispatch_q_qdot_tau_taudot_data(states: MX.sym, controls: MX.sym, nlp) -> tuple:
+        """
+        Extracting q, qdot, tau, taudot, from states and controls, assuming state, state, state and control,
+         respectively.
+
+        Parameters
+        ----------
+        states: MX.sym
+            The state of the system
+        controls: MX.sym
+            The controls of the system
+        nlp: NonLinearProgram
+            The definition of the system
+
+        Returns
+        ----------
+        MX.sym
+            q, the generalized coordinates
+        MX.sym
+            qdot, the generalized velocities
+        MX.sym
+            tau, the generalized torques
+        MX.sym
+            taudot, the generalized torque derivatives
+        """
+
+        nq = nlp.mapping["q"].to_first.len
+        q = nlp.mapping["q"].to_second.map(states[:nq])
+        qdot = nlp.mapping["qdot"].to_second.map(states[nq:2*nq])
+        tau = nlp.mapping["tau"].to_second.map(states[2*nq:])
```

Comment:
Please note that nq is not necessarily equal to nqdot and ntau. The function should get them all separately

Reply:
How is it possible ?

Reply:
If a dof from a segment is a quaternion, then nq for that segment is 4, but n for the rest is 3 as velocities and accelerations lies in a perpendicular dimension at the tip of the quaternion vector